### PR TITLE
Add invalidating mod_auth_session capabilities

### PR DESCRIPTION
**Pull request by gurtzoo**

### CODE REVIEW

**Comment by jhrozek** [comment]
So this is "a client side logout" ?

I think you might want to call `am_check_url` to prevent being redirected to somewhere you wouldn't want to be redirected..

File: README.md [hunk 1]
```diff
@@ -744,6 +744,12 @@ This will return the user to "https://www.example.org/logged_out.html"
 after the logout operation has completed.
 
 
+## Invalidating session
+It is possible to invalidate the current mod_auth_mellon session, without calling SLO. 
+The mod_auth_mellon cookie session will be invalidate and session will be removed from mod_auth_mellon cache.
```
Comment (simo5) [inline README.md]: invalidate and session -> invalidated and the session
from mod_auth_mellon -> from the mod_auth_mellon

File: README.md [hunk 2]
```diff
@@ -744,6 +744,12 @@ This will return the user to "https://www.example.org/logged_out.html"
...
+The invalidate the session, just call the endpoint "<endpoint path>/invalidate".
```
Comment (simo5) [inline README.md]: The invalidate -> To invalidate

File: auth_mellon.h [hunk 3]
```diff
@@ -476,6 +476,7 @@ am_cache_entry_t *am_get_request_session_by_nameid(request_rec *r,
 am_cache_entry_t *am_new_request_session(request_rec *r);
 void am_release_request_session(request_rec *r, am_cache_entry_t *session);
 void am_delete_request_session(request_rec *r, am_cache_entry_t *session);
+void am_invalidate_request_session(request_rec *r);
```
Comment (simo5) [inline auth_mellon.h]: The previous two function pass the session as argument, why not follow the same style here ?

File: auth_mellon_handler.c [hunk 4]
```diff
@@ -807,6 +807,54 @@ static int am_handle_logout_request(request_rec *r,
...
+static int am_handle_invalidate_request(request_rec *r)
+{
+    int rc;
+    char *return_to;
+    am_cache_entry_t *session = am_get_request_session(r);
+
+    am_diag_printf(r, "enter function %s\n", __func__);
+    am_diag_log_cache_entry(r, 0, session, "%s\n", __func__);
+
+    if(session != NULL) {
```
Comment (simo5) [inline auth_mellon_handler.c]: nits: space before parens
if(session-> if (session

File: auth_mellon_handler.c [hunk 5]
```diff
@@ -807,6 +807,54 @@ static int am_handle_logout_request(request_rec *r,
...
+        am_delete_request_session(r, session);
+    }
+
+    return_to = am_extract_query_parameter(r->pool, r->args, "ReturnTo");
+
+    if(return_to == NULL) {
```
Comment (simo5) [inline auth_mellon_handler.c]: same nits

File: auth_mellon_handler.c [hunk 6]
```diff
@@ -807,6 +807,54 @@ static int am_handle_logout_request(request_rec *r,
...
+        AM_LOG_RERROR(APLOG_MARK, APLOG_ERR, 0, r,
+                      "No ReturnTo parameter provided for invalidate handler.");
+        return HTTP_BAD_REQUEST;
```
Comment (simo5) [inline auth_mellon_handler.c]: if the request is malformed, should you still delete the session anyway ?

Comment (simo5) [reply auth_mellon_handler.c]: if yes then I would not consider the req malformed and simply redirect to the referer or missing that the main mod_mellon page ?

File: auth_mellon_handler.c [hunk 7]
```diff
@@ -807,6 +807,54 @@ static int am_handle_logout_request(request_rec *r,
...
+    }
+
+    rc = am_urldecode(return_to);
+    if(rc != OK) {
```
Comment (simo5) [inline auth_mellon_handler.c]: same nit

File: auth_mellon_handler.c [hunk 8]
```diff
@@ -807,6 +807,54 @@ static int am_handle_logout_request(request_rec *r,
...
+        AM_LOG_RERROR(APLOG_MARK, APLOG_ERR, rc, r,
+                      "Could not urldecode ReturnTo value in invalidate"
+                      " response.");
+        return HTTP_BAD_REQUEST;
```
Comment (simo5) [inline auth_mellon_handler.c]: same comment as above, feels weird to me to do the action first and then check if parameters are correct

File: auth_mellon_handler.c [hunk 9]
```diff
@@ -807,6 +807,54 @@ static int am_handle_logout_request(request_rec *r,
...
+    int rc;
+    char *return_to;
+    am_cache_entry_t *session = am_get_request_session(r);
+
+    am_diag_printf(r, "enter function %s\n", __func__);
+    am_diag_log_cache_entry(r, 0, session, "%s\n", __func__);
+
+    if(session != NULL) {
+        am_delete_request_session(r, session);
+    }
```
Comment (simo5) [inline auth_mellon_handler.c]: if the session is NULL, should this give back an error ?

Comment (gurtzoo) [reply auth_mellon_handler.c]: yes i will add that

File: auth_mellon_handler.c (hunk 8 above)
Comment (gurtzoo) [reply auth_mellon_handler.c]: I have base the code on the "logout" code.
But i think you are right, i will invalidate at the end of the validations.

**Comment by gurtzoo** [comment]
Yes, the "invalidate" is like a client side logout, by deleting the session (cache, and cookie).

File: auth_mellon_handler.c (hunk 4 above)
Comment (jhrozek) [reply auth_mellon_handler.c]: In all honesty, the mellon code is not consistent with the spaces after keywords, so it gets easy to copy the same style issue from elsewhere.
But I do agree that the spaces are better readable.

File: auth_mellon.h (hunk 3 above)
Comment (jhrozek) [reply auth_mellon.h]: Hmm, more importantly I don't see the function being neither defined nor used anywhere?

**Comment by gurtzoo** [comment]
Have you review my last commits (4 nov)?

File: auth_mellon_handler.c [hunk 10]
```diff
@@ -1139,6 +1197,25 @@ static int am_handle_logout(request_rec *r)
...
+ *
+ * Returns:
+ *  OK on success, or an error if any of the steps fail.
+ */
+static int am_handle_invalidate(request_rec *r)
+{
+    LassoServer *server;
+
+    server = am_get_lasso_server(r);
+    if(server == NULL) {
```
Comment (simo5) [inline auth_mellon_handler.c]: still style nit, space between if and (

File: auth_mellon.h [hunk 11]
```diff
@@ -476,6 +476,7 @@ am_cache_entry_t *am_get_request_session_by_nameid(request_rec *r,
 am_cache_entry_t *am_new_request_session(request_rec *r);
 void am_release_request_session(request_rec *r, am_cache_entry_t *session);
 void am_delete_request_session(request_rec *r, am_cache_entry_t *session);
+void am_invalidate_request_session(request_rec *r, am_cache_entry_t *session);
```
Comment (simo5) [inline auth_mellon.h]: You changed this function declaration, but there is still no implementation of this function, what's the point ??

**Comment by gurtzoo** [comment]
Hi, i have done the latest changes.

File: README.md [hunk 12]
```diff
@@ -744,6 +744,12 @@ This will return the user to "https://www.example.org/logged_out.html"
 after the logout operation has completed.
 
 
+## Invalidating session
+It is possible to invalidate the current mod_auth_mellon session, without calling SLO. 
+The mod_auth_mellon cookie session will be invalidated and the session will be removed from the mod_auth_mellon cache.
+To invalidate the session, just call the endpoint "<endpoint path>/invalidate".
+
```
Comment (simo5) [inline README.md]: Please add a warning that is session is invalidated then SLO will become impossible for that session. sorry for not asking earlier it just occurred to me that this is a side effect that users need to be aware of.
I am also wondering if this feature should be always allowed or should rather require a configure knob. I can see how some admins may not want to allow users to drop sessions without performing a full SLO.

**Comment by simo5** [comment]
Also please squash all commits into one

File: README.md (hunk 12 above)
Comment (gurtzoo) [reply README.md]: Yes i will add the warning.
Do you think about a toggle switch like "enabledInvalidateEndpoint"?

Comment (simo5) [reply README.md]: yes, sounds reasonable.

**Comment by gurtzoo** [comment]
Hi, i have add the comments and also introduce a variable: MellonEnabledInvalidateSessionEndpoint to enabled the endpoint, default "off".
Commits as has been squashed.

**Comment by simo5** [comment]
I still see 6 commits, the comments looks good though.

**Comment by gurtzoo** [comment]
hoping this time the squash is really done! ;-)

**Pull request merged**